### PR TITLE
feat(manager/gradle): enable support for concatenated properties

### LINKED ISSUE #8728: Unsupported dependency in gradle
**What would you like Renovate to be able to do?**
Hi, i have detected an unsupported dependency style for the gradle and gradle-lite manager.

The style looks like this
```
    testImplementation 'com.github.gmazzo:okhttp-mock:' + okhttpMockVersion
```

Maybe its possible to support this as well?

**Did you already have any implementation ideas?**

I guess the parsing must be adjusted?

Thanks
eloo

### PR DESCRIPTION
**Pull request by Churro**
## Changes

Adds parsing support for up to 32 properties chained via `+` operator. This enables advanced string combinations like:
```groovy
maven {
  url = repobase + "/${name}"
}

dependencies {
    classpath("io.jsonwebtoken:" + project.property("somepart") + "-api:" + project.ext["jjwt_version"])
}
```

- Parsing of regular dependency strings isn't changed with this PR. In case further properties are found _after_ `+`, they are memorized as additional tokens. This is already done the same way for template strings like `"foobar${baz}"`. String combinations, such as `"foo" + "bar" + "${baz}"` can be processed with the same handler methods.
- New tests partly include `fileReplacePosition` to validate the handling that was added to `qVariableAccessIdentifier`. It is responsible to ensure that `"some" + foo.bar` isn't mistakenly folded into `some.foo.bar`

// Update: I've now enabled `+` for `version` too. At first, the intention was not to do this, since any concatenated version string could not be updated properly anyway. Upon thinking it further, I believe it makes sense to have it enabled for more correct version extractions. E.g. a dep specified with `version: kafkaVersion + "-ce"` should rather be extracted with `unknown-version` than looked up with the value of `kafkaVersion` only, which will eventually produce no or wrong lookup results.

## Context

* Closes https://github.com/renovatebot/renovate/issues/8728

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Test repo: https://github.com/renovate-demo/8728-concat

### CODE REVIEW

**Review by zharinov — DISMISSED** [review]

**Review by rarkins — DISMISSED** [review]

**Comment by Churro** [comment]
I'd suggest to wait for https://github.com/renovatebot/renovate/pull/19672. There's one line in this PR that will need adjustment to work with `good-enough-parser:v1.1.22`

**Comment by zharinov** [comment]
Oh, okay, thanks

File: lib/modules/manager/gradle/parser.spec.ts [hunk 1]
```diff
@@ -351,6 +351,27 @@ describe('modules/manager/gradle/parser', () => {
...
+        ${'baz = "1.2.3"'}                   | ${'"foo:bar:" + property("baz")'} | ${{ depName: 'foo:bar', currentValue: '1.2.3', managerData: { fileReplacePosition: 7 } }}
+        ${'baz = "1.2.3"'}                   | ${'"foo:bar:" + baz + "456"'}     | ${{ depName: 'foo:bar', currentValue: '1.2.3456', skipReason: 'contains-variable' }}
+        ${'baz = "1.2.3"'}                   | ${'"foo:bar:" + baz + "@zip"'}    | ${{ depName: 'foo:bar', currentValue: '1.2.3', dataType: 'zip', managerData: { fileReplacePosition: 7 } }}
+        ${'foo.bar = "bar:"; baz = "1.2.3"'} | ${'"foo:" + foo.bar + "${baz}"'}  | ${{ depName: 'foo:bar', currentValue: '1.2.3', managerData: { fileReplacePosition: 25 } }}
+        ${'foo.bar = "bar"; baz = "1.2.3"'}  | ${'"foo:bar_${foo.bar}:" + baz'}  | ${{ depName: 'foo:bar_bar', currentValue: '1.2.3', managerData: { fileReplacePosition: 24 } }}
+      `('$def | $str', ({ def, str, output }) => {
+        const { deps } = parseGradle([def, str].join('\n'));
+        expect(deps).toMatchObject([output].filter(Boolean));
```
Comment (viceice) [inline lib/modules/manager/gradle/parser.spec.ts:371]: please use `is.truthy` instead . why is that filter needed?

File: lib/modules/manager/gradle/parser/common.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
-import { lexer, query as q } from 'good-enough-parser';
+import { lexer, parser, query as q } from 'good-enough-parser';
+import type { SeqBuilder } from 'good-enough-parser/dist/cjs/query/builder';
```
Comment (viceice) [inline lib/modules/manager/gradle/parser/common.ts:2]: @zharinov type should be exported

File: lib/modules/manager/gradle/parser.spec.ts (hunk 1 above)
Comment (Churro) [reply lib/modules/manager/gradle/parser.spec.ts]: Good catch, not needed at all here. I've changed the other occurrences to use `is.truthy`

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/gradle/parser/common.ts (hunk 2 above)
Comment (zharinov) [reply lib/modules/manager/gradle/parser/common.ts:2]: No, these classes are implementation detail. Need another way 🤔

Comment (zharinov) [reply lib/modules/manager/gradle/parser/common.ts:2]: PR-for-PR: https://github.com/Churro/renovate/pull/1

Comment (Churro) [reply lib/modules/manager/gradle/parser/common.ts]: Thanks!

**Review by zharinov — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**